homeassistant/helpers/entity.py

```diff
@@ -453,26 +453,35 @@ async def async_device_update(self, warning: bool = True) -> None:
         if self.parallel_updates:
             await self.parallel_updates.acquire()
 
-        assert self.hass is not None
-        if warning:
-            update_warn = self.hass.loop.call_later(
-                SLOW_UPDATE_WARNING,
-                _LOGGER.warning,
-                "Update of %s is taking over %s seconds",
-                self.entity_id,
-                SLOW_UPDATE_WARNING,
-            )
-
         try:
             # pylint: disable=no-member
             if hasattr(self, "async_update"):
-                await self.async_update()  # type: ignore
+                task = self.hass.async_create_task(self.async_update())  # type: ignore
             elif hasattr(self, "update"):
-                await self.hass.async_add_executor_job(self.update)  # type: ignore
+                task = self.hass.async_add_executor_job(self.update)  # type: ignore
+            else:
+                return
+
+            if not warning:
+                await task
+                return
+
+            finished, _ = await asyncio.wait([task], timeout=SLOW_UPDATE_WARNING)
```

Comment:
Why not wait_for?

Reply:
With `wait_for`, I had to use `asyncio.shield` and the implementation got a bit complex.

Reply:
Oh right you want it to continue